File: 0234-palindrome-linked-list/0234-palindrome-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

/*
func isPalindrome(head *ListNode) bool {
    candidate := []int{}
    cur := head
    
    for cur != nil {
        candidate = append(candidate, cur.Val)
        cur = cur.Next
    }
    
    for i := 0; i < len(candidate)/2; i++ {
        if candidate[i] != candidate[len(candidate)-1-i] {
            return false
        }
    }
    
    return true
}
*/

func isPalindrome(head *ListNode) bool {
    var prev *ListNode
    slow := head
    fast := slow
    
    for fast != nil && fast.Next != nil {
        // advance 'fast' at double the rate
        fast = fast.Next
        fast = fast.Next
        
        // reverse the first half of the list while advancing
        slowNext := slow.Next
        slow.Next = prev
        prev = slow
        slow = slowNext
    }
    
    // advance past midpoint for odd-length lists
    if fast != nil {
        slow = slow.Next
    }
    
    firstHalf := prev
    secondHalf := slow
    
    for firstHalf != nil {
        if firstHalf.Val != secondHalf.Val {
            return false
        }
        
        firstHalf = firstHalf.Next
        secondHalf = secondHalf.Next
    }
    
    return true
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
"""
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        candidate = []
        cur = head
        
        while cur != None:
            candidate.append(cur.val)
            cur = cur.next
            
        for i in range(len(candidate)//2):
            if candidate[i] != candidate[len(candidate)-1-i]:
                return False
            
        return True    
"""

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        prev = None
        slow = head
        fast = slow
        
        while fast and fast.next:
            fast = fast.next.next
            
            slow_next = slow.next
            slow.next = prev
            prev = slow
            slow = slow_next
        
        if fast:
            slow = slow.next
            
        first_half = prev
        second_half = slow
        
        while first_half and second_half:
            if first_half.val != second_half.val:
                return False
            first_half = first_half.next
            second_half = second_half.next
            
        return True    
        
            